testbed/tests/window/test_window.py

```diff
@@ -571,6 +796,145 @@ async def test_full_screen(second_window, second_window_probe):
+        await second_window_probe.wait_for_window(
+            "Secondary window is not in presentation mode", full_screen=True
+        )
+        assert second_window_probe.get_window_state() != WindowState.PRESENTATION
+        assert second_window_probe.is_resizable
+        assert second_window_probe.presentation_content_size == initial_content_size
+
+        second_window.state = WindowState.NORMAL
+        await second_window_probe.wait_for_window(
+            "Secondary window is still not in presentation mode", full_screen=True
+        )
+        assert second_window_probe.get_window_state() != WindowState.PRESENTATION
+        assert second_window_probe.presentation_content_size == initial_content_size
+
+    @pytest.mark.parametrize(
+        "initial_state, final_state",
+        [
+            # Direct switch from MINIMIZED:
+            (WindowState.MINIMIZED, WindowState.MAXIMIZED),
+            (WindowState.MINIMIZED, WindowState.FULLSCREEN),
+            (WindowState.MINIMIZED, WindowState.PRESENTATION),
+            # Direct switch from MAXIMIZED:
+            (WindowState.MAXIMIZED, WindowState.MINIMIZED),
+            (WindowState.MAXIMIZED, WindowState.FULLSCREEN),
+            (WindowState.MAXIMIZED, WindowState.PRESENTATION),
+            # Direct switch from FULLSCREEN:
+            (WindowState.FULLSCREEN, WindowState.MINIMIZED),
+            (WindowState.FULLSCREEN, WindowState.MAXIMIZED),
+            (WindowState.FULLSCREEN, WindowState.PRESENTATION),
+            # Direct switch from PRESENTATION:
+            (WindowState.PRESENTATION, WindowState.MINIMIZED),
+            (WindowState.PRESENTATION, WindowState.MAXIMIZED),
+            (WindowState.PRESENTATION, WindowState.FULLSCREEN),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "second_window_class, second_window_kwargs",
+        [
+            (
+                toga.MainWindow,
+                dict(title="Secondary Window", position=(200, 150)),
+            )
+        ],
+    )
+    async def test_window_state_direct_change(
+        app, initial_state, final_state, second_window, second_window_probe
+    ):
+        if (
+            WindowState.MINIMIZED in {initial_state, final_state}
+            and not second_window_probe.supports_minimize
+        ):
+            pytest.xfail(
+                "This backend doesn't reliably support minimized window state."
+            )
+        second_window.toolbar.add(app.cmd1)
+        second_window.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
+        second_window.show()
+
+        # A longer delay to allow for genie animations
+        await second_window_probe.wait_for_window(
+            "Secondary window is visible", full_screen=True
+        )
+
+        assert second_window_probe.get_window_state() == WindowState.NORMAL
+        assert second_window_probe.get_window_state() != initial_state
+        assert second_window_probe.get_window_state() != final_state
```

Comment:
It's also worth pointing out that assertions like this are *massively* over specified. You don't need to check something is equal to one value, then not equal to all other values. If `window state == NORMAL`, then, by definition, it isn't the initial state, and it isn't the final state. This "overspecified assertion" pattern exists throughout this test file.

Reply:
I have removed these type of assertions.